identity/src/wallet.rs

```diff
@@ -0,0 +1,466 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::{convert::TryFrom, str::FromStr};
+
+use ahash::{HashMap, HashMapExt};
+use fvm_shared::{
+    address::Address,
+    crypto::signature::{Signature, SignatureType},
+};
+use serde::{Deserialize, Serialize};
+
+use crate::{errors::Error, wallet_helpers, KeyInfo, KeyStore};
+
+/// A key, this contains a `KeyInfo`, an address, and a public key.
+#[derive(Clone, PartialEq, Debug, Eq, Serialize, Deserialize)]
+pub struct Key {
+    pub key_info: KeyInfo,
+    // Vec<u8> is used because The public keys for BLS and SECP256K1 are not of the same type
+    pub public_key: Vec<u8>,
+    pub address: Address,
+}
+
+impl TryFrom<KeyInfo> for Key {
+    type Error = crate::errors::Error;
+
+    fn try_from(key_info: KeyInfo) -> Result<Self, Self::Error> {
+        let public_key = wallet_helpers::to_public(*key_info.key_type(), key_info.private_key())?;
+        let address = wallet_helpers::new_address(*key_info.key_type(), &public_key)?;
+        Ok(Key {
+            key_info,
+            public_key,
+            address,
+        })
+    }
+}
+
+// This is a Wallet, it contains 2 HashMaps:
+// - keys which is a HashMap of Keys resolved by their Address
+// - keystore which is a HashMap of KeyInfos resolved by their Address
+/// A wallet is a collection of private keys with optional persistence and
+/// optional encryption.
+#[derive(Clone, PartialEq, Debug, Eq)]
+pub struct Wallet {
+    keys: HashMap<Address, Key>,
+    keystore: KeyStore,
+}
+
+impl Wallet {
+    /// Return a new wallet with a given `KeyStore`
+    pub fn new(keystore: KeyStore) -> Self {
+        Wallet {
+            keys: HashMap::new(),
+            keystore,
+        }
+    }
+
+    /// Return a wallet from a given amount of keys.
+    pub fn new_from_keys(keystore: KeyStore, key_vec: impl IntoIterator<Item = Key>) -> Self {
+        let mut keys: HashMap<Address, Key> = HashMap::new();
+        for item in key_vec.into_iter() {
+            keys.insert(item.address, item);
+        }
+        Wallet { keys, keystore }
+    }
+
+    // If this key does not exist in the keys hashmap, check if this key is in
+    // the keystore, if it is, then add it to keys, otherwise return Error
+    /// Return the key that is resolved by a given address,
+    pub fn find_key(&mut self, addr: &Address) -> Result<Key, Error> {
```

Comment:
This api is causing some issues downstream. From the function name, it's a `getter`, but it's taking a `&mut self`, this makes anyone who wants to find a key that holds the `wallet` need to use `&mut self`. And it complicates things downstream. If possible, we should remove the `&mut`.